chamelon/compat.jst.ml

```diff
@@ -114,32 +170,50 @@ let view_texp (e : expression_desc) =
   | Texp_construct (name, desc, args, mode) ->
       Texp_construct (name, desc, args, mode)
   | Texp_tuple (args, mode) -> Texp_tuple (args, mode)
-  | Texp_function
-      {
-        arg_label;
-        param;
-        cases;
-        partial;
-        arg_mode;
-        alloc_mode;
-        region;
-        curry;
-        warnings;
-        arg_sort;
-        ret_sort;
-      } ->
-      Texp_function
-        ( { arg_label; param; cases },
-          {
-            partial;
-            arg_mode;
-            alloc_mode;
-            region;
-            curry;
-            warnings;
-            arg_sort;
-            ret_sort;
-          } )
+  | Texp_function { params; body; alloc_mode; region; ret_sort } -> (
+      let exception Unviewable_function in
+      try
+        let params =
+          List.map
+            (fun param ->
+              match param.fp_kind with
+              | Tparam_optional_default _ -> raise Unviewable_function
```

Comment:
I'm slightly worried about this, because it can result is a `Texp_function` sometimes being seen as `O (Texp_function _)` instead as `Texp_function _`, preventing all minimizers that should apply on functions to work there (still, this is better than failing altogether: at least we can run the minimizers on subexpressions).

The best thing to do would arguably to add the possibility of optional params in `texp_function_param` (and simply never generate them with upstream compatibility, while expanding them when going the other way); but it might require changes in the various minimizers to match.
Some slight reassurances: we don't have to handler every single pathological case; it is enough to have something "best effort" that works in most cases, so we might not have to worry about replicating the logic of `push_defaults` or the exact sequencing order. However, it would be good to have the minimizer still be able to inline or remove unused arguments of functions with optional arguments, and this prevents it, as it won't be recognized as a function.

Reply:
I've now changed it as you suggested, adding `optional_default` to `texp_function_param`. It didn't seem to require changes in the minimizers, and some manual test cases look fine. (I'll add a comment to the main PR thread with a small example.)

Reply:
Thanks, it looks good now! I was a bit worried about `compat.upstream.ml` failing on a param with an optional default, but it looks like this cannot happen, so all good.